tensorflow-frameworks/tensorflow-data/src/main/java/org/tensorflow/data/DatasetIterator.java

```diff
@@ -0,0 +1,214 @@
+package org.tensorflow.data;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.tools.Shape;
+
+import java.util.List;
+
+/**
+ * Represents the state of an iteration through
+ * a tf.data Datset.
+ * <p>
+ * Example: Iteration in graph mode.
+ *
+ * <pre>{@code
+ *  // Create input tensors
+ *  Operand<?> XTensor = tf.constant( ... );
+ *  Operand<?> yTensor = tf.constant( ... );
+ *
+ *
+ *  Dataset dataset = Dataset
+ *          .fromTensorSlices(XTensor, yTensor);
+ *          .batch(BATCH_SIZE);
+ *
+ *  DatasetIterator iterator = dataset.makeIterator();
+ *  List<Output<?>> components = iterator.getNext();
+ *  Operand<?> XBatch = components.get(0);
+ *  Operand<?> yBatch = components.get(1);
+ *
+ *  // Build a TensorFlow graph that does something on each element.
+ *  loss = computeModelLoss(X, y);
+ *
+ *  optimizer = ... // create an optimizer
+ *  trainOp = optimizer.minimize(loss);
+ *
+ *  try (Session session = new Session(graph) {
+ *      try {
+ *          session.run(trainOp);
+ *          ...
+ *      } catch (IndexOutOfBoundsException e) {
+ *          System.out.println("finished iterating.");
+ *          break;
+ *      }
+ *  }
+ *
+ * }</pre>
+ * <p>
+ * Example: Iteration in eager mode.
+ *
+ * <pre>{@code
+ *  // Create input tensors
+ *  Operand<?> XTensor = tf.constant( ... );
+ *  Operand<?> yTensor = tf.constant( ... );
+ *
+ *  int BATCH_SIZE = ...
+ *
+ *  Dataset dataset = Dataset
+ *          .fromTensorSlices(XTensor, yTensor)
+ *          .batch(BATCH_SIZE);
+ *  DatasetIterator iterator = dataset.makeIterator();
+ *
+ *  Optimizer optimizer = ... // create an optimizer
+ *
+ *  for (List<Output<?>> components : dataset) {
+ *      Operand<?> XBatch = components.get(0);
+ *      Operand<?> yBatch = components.get(1);
+ *
+ *      loss = computeModelLoss(X, y);
+ *      trainOp = optimizer.minimize(loss);
+ *  }
+ * }</pre>
+ */
+class DatasetIterator {
```

Comment:
This class is package private, yet returned by a public method on `Dataset`. Should it be `public`?

Reply:
Yup good catch, updated!